File: _BHCVRP_Python_Version/cujae_inf_citi_om/data/BusStop.py
from data.Location import Location
from exceptions.BusStopCapacityException import BusStopCapacityException

# Clase que modela los datos de una parada en un SBRP

class BusStop:
    def __init__(
        self,
        id_bus_stop=None,
        capacity_bus_stop=None,
        location_bus_stop: Location = None,
        list_customers = None
    ):
        if (
            id_bus_stop is not None
            and capacity_bus_stop is not None
            and location_bus_stop is not None
            and list_customers is not None
        ):
            # Constructor con tres argumentos
            self._id_bus_stop = id_bus_stop
            self._capacity_bus_stop = capacity_bus_stop
            self._location_bus_stop = location_bus_stop
            self._list_customers = list_customers

        else:
            # Constructor sin argumentos (o con argumentos predeterminados)
            self._id_bus_stop = 0
            self._capacity_bus_stop = 0.0
            self._location_bus_stop = None  # O crea una nueva instancia de Location con valores predeterminados si es necesario
            self._list_customers = []

    def get_id_bus_stop(self):
        return self._id_bus_stop

    def set_id_bus_stop(self, id_bus_stop):
        self._id_bus_stop = id_bus_stop

    def get_capacity_bus_stop(self):
        return self._capacity_bus_stop

    def set_capacity_bus_stop(self, capacity_bus_stop):
        if capacity_bus_stop > 0:
            self._capacity_bus_stop = capacity_bus_stop
        else:
            raise BusStopCapacityException("La capacidad de la parada debe ser mayor que cero")

    def get_location_bus_stop(self):
        return self._location_bus_stop

    def set_location_bus_stop(self, location_bus_stop: Location):
        self._location_bus_stop = location_bus_stop

    def get_list_customers(self):
        return self._list_customers

    def set_list_customers(self, list_customers):
        self._list_customers = list_customers

    def has_capacity(self):
        return self._capacity_bus_stop > self._list_customers.__len__()

    def insert_customer(self, id_customer):
        if self.has_capacity():
            self._list_customers.append(id_customer)

        return self.has_capacity()

    def increase_capacity_bus_stop(self, increase):
        if increase > 0:
            self._capacity_bus_stop += increase

    def decrease_capacity_bus_stop(self, decrease):
        if decrease > 0 and (self._capacity_bus_stop - decrease > 0):
            self._capacity_bus_stop -= decrease
        else:
            raise BusStopCapacityException("La capacidad de la parada debe ser mayor que cero")

    def get_coordinates(self):
        return self._location_bus_stop.get_axis_x() + " " + self._location_bus_stop.get_axis_y()

File: BHCVRP_Java_Version/src/cujae/inf/citi/om/data/BusStop.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package cujae.inf.citi.om.data;

import java.util.ArrayList;
import java.util.stream.Collectors;

/**
 *
 * @author nanda
 */

//Clase que modela los datos de una parada en un SBRP.

public class BusStop {
    private String idBusStop;
    private Location locationBusStop;
    private int capacityBusStop;
    private ArrayList<String> listCustomers;
    
    public BusStop(){
        super();
    }
    
    public BusStop(String idBusStop, int capacity, Location locationBus) {
        setIdBusStop(idBusStop);
        setLocationBusStop(locationBus);
        setCapacityBusStop(capacity);

        this.listCustomers = new ArrayList<String>();
    }
    
    public BusStop(int idBusStop, int capacity, Location locationBus, ArrayList<Integer> listCustomers){
        this.idBusStop = String.valueOf(idBusStop);
        setLocationBusStop(locationBus);
        setCapacityBusStop(capacity);

        this.listCustomers = (ArrayList<String>) listCustomers.stream().map(String::valueOf).collect(Collectors.toList());
    }
    
    public String getIdBusStop(){
        return idBusStop;
    }
    
    public void setIdBusStop(String idBusStop) throws IllegalArgumentException{
        if(!idBusStop.equals(null)){
            this.idBusStop = idBusStop;
        }
        else
            throw new IllegalArgumentException();     
    }
    
    public Location getLocationBusStop(){
        return locationBusStop;
    }
    
    public void setLocationBusStop(Location locationBusStop) throws IllegalArgumentException{
        if(!locationBusStop.equals(null)){
            this.locationBusStop = locationBusStop;
        }
        else
            throw new IllegalArgumentException();            
    }    
    
    public int getCapacityBusStop(){
        return capacityBusStop;
    } 
    
    public void setCapacityBusStop(int capacityBusStop){
        if(capacityBusStop > 0){
            this.capacityBusStop = capacityBusStop;
        }
    }
    
     public ArrayList<String> getListCustomers(){
         return listCustomers;
     }
     
     //Método para determinar si hay capacidad en la parada.
     public boolean hasCapacity(){
         return capacityBusStop > listCustomers.size();
     }
     
     //Método para saber si se puede insertar un nuevo pasajero.
     public boolean insertCustomer(String idCustomer){
         if(hasCapacity()){
             listCustomers.add(idCustomer);
         }
         
         return hasCapacity();
     }
     
     //Método para aumentar la capacidad de la parada.
     public void increaseCapacityBusStop(int increase){
         if(increase > 0){
             this.capacityBusStop += increase;
         }
     }
     
     //Método para decrementar la capacidad de la parada.
     public void decreaseCapacityBusStop(int decrease){
         if(decrease > 0 && (capacityBusStop - decrease > 0) ){
            this.capacityBusStop -= decrease;
         }
            
     }
     
     public String get_coordinates(){
    
        return this.locationBusStop.getAxisX() + " " + this.locationBusStop.getAxisY();
        
    }
     
}
